### PR TITLE
Binary decoder for stake pool registrations.

### PR DESCRIPTION
**Pull request by rvl**
Relates to #1089 

# Overview

 - Adds a binary decoder for stake pool registration certificates.
 - Use jcli to generate golden test data for pool registration fragments.

### CODE REVIEW

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 1]
```diff
@@ -482,6 +486,32 @@ data MkFragment
...
+    :: Get (PoolId, Tx)
+getPoolRetirement = error "retirement not implemented"
+
+getPoolRegistration
+    :: Get (PoolId, [PoolOwner], Tx)
+getStakeDelegation = do
+    label "getPoolRegistation" $ do
+        _serial <- getByteString 16
+        _perms <- getByteString 8
+        _keys <- getByteString 256 -- fixme: check POOL-KEYS = VRF-PUBLICKEY KES-PUBLICKEY
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs]: **[suggestion]**

Use [skip](https://hackage.haskell.org/package/binary-0.10.0.0/docs/Data-Binary-Get.html#v:skip) with a clear comment about what's being skipped.

**Why**
Most of these doesn't really deserialize to `ByteString` and I suppose, this is really what you intended to write ^^"

File: default.nix [hunk 2]
```diff
@@ -64,6 +64,7 @@ let
         ++ [ jormungandr jormungandr-cli
              pkgs.pkgconfig pkgs.sqlite-interactive ];
       meta.platforms = pkgs.lib.platforms.unix;
+      shellHook = "unset NIX_BUILD_TOP";
```
Comment (KtorZ) [inline default.nix]: **[question]**

:thinking: What is this ?

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 3]
```diff
@@ -491,7 +491,8 @@ isSubrangeOf r1 r2 =
                                   Stake Pools
 -------------------------------------------------------------------------------}
 
--- | Represents a stake pool.
+-- | Identifies a stake pool. A 'PoolId' is the blake2b-256 hash of the stake
+-- pool registration certificate.
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs]: **[remark]**

Yes and no. This is true for Jörmungandr, but not generally true of PoolIds. We have to be careful about these things.

File: lib/core/src/Cardano/Wallet/Unsafe.hs [hunk 4]
```diff
@@ -110,3 +123,20 @@ unsafeDeserialiseCbor decoder bytes = either
...
+
+-- | Get the data part of a bech32-encoded string, ignoring the human-readable part.
+unsafeBech32Decode :: HasCallStack => Text -> BL.ByteString
+unsafeBech32Decode txt = case Bech32.decodeLenient txt of
+    Right (_hrp, dp) -> maybe (bomb "missing data part")
+        BL.fromStrict (Bech32.dataPartToBytes dp)
+    Left e -> bomb (show e)
+  where
+    bomb msg = error $ "Could not decode bech32 string " ++ show txt
+        ++ " because " ++ msg
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Unsafe.hs:142]: :+1:

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 5]
```diff
@@ -417,11 +420,15 @@ getStakeDelegation tid = do
...
-        poolId <- getByteString 32
-        pure $ Just $ DlgFull $ PoolId poolId
+        poolId <- getPoolId
+        pure $ Just $ DlgFull poolId
     getStakeDelegationRatio =
         MaybeT (pure Nothing)
 
+-- | Corresponds to @POOL-ID@ in the specification.
+getPoolId :: Get PoolId
+getPoolId = PoolId <$> getByteString 32
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:430]: :+1:

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 6]
```diff
@@ -494,6 +501,85 @@ data MkFragment
...
+    getCert = label "REGISTRATION-CERT" $ do
+        -- A random value, for user purpose similar to a UUID.  it may not be
+        -- unique over a blockchain, so shouldn't be used a unique identifier
+        _serial <- label "POOL-SERIAL" $ skip 16
+        -- Beginning of validity for this pool. A time in seconds since genesis.
+        -- This is used to keep track of the period of the expected key and the
+        -- expiry
+        startTime <- label "TIME-SINCE-EPOCH0" getWord64be
+        -- Permission system for this pool.
+        label "POOL-PERMISSIONS" $ skip 8
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:529]: :+1:

**Review by KtorZ — APPROVED** [review]

File: default.nix (hunk 2 above)
Comment (rvl) [reply default.nix]: It lets `Test.Utils.Paths.getTestData` work under `nix-shell`, but I will put it in another PR.

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 3 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs]: OK

**Comment by rvl** [comment]
Forgot to hlint...

bors r+

**Pull request merged**